### PR TITLE
Datajson v1.0

### PR DESCRIPTION
**Pull request by regit**
Indicator of Compromises (IOCs) are a key element in Security Operating Center. Dataset
have been a huge step in getting alert on IOCs from Suricata. But produced alerts are
lacking contextualization. For example, if a IOC management software has a list of host
names, they will be linked to different threat actors but at ingestion in Suricata they will be
a simple list of strings. This list will be used via a rule like

```
alert tls any any -> any any (msg:"IOC hostname on TLS"; tls.sni; dataset:isset,hostname.lst,...; sid:1664;)
```

 With this an alert will have a subject without information and a mapping will have to be done
 at posteriori to see which IOC has hit. The pseudo algorithm to run is:

  - Intercept the signature 1664
  - Extract the tls.sni
  - Check the tls.sni value in the IOC management software

This works but it is not optimal as correlation and external processing has to be done for all the matches.

To fix this issue, we need to be able to ingest the IOCs without loosing the contextual
information contained in the IOC management software.

Datajson is a proposed implementation that addresses this issue. Instead of injecting into
Suricata the value list we can attach to each value a JSON object that will end up into
the alert output.

The following example is alert on source and destination IP in dataset:

```
alert tls $HOME_NET any -> any any (msg:"Test dataset";
                              ip.src; datajson:isset,ip4list,type ipv6,load ip4-json.lst,key inventory; \
                              ip.dst; datajson:isset,actors_ip,type ipv6, load bad-json.lst, key bad_actors; sid:1;)
```

In ip4-json.lst, we have data from inventory:

```
10.7.5.5,{"user":"vjulien","rank": 1}
```

In bad-json.lst, we have data from the IOC management software:

```
185.117.73.76,["Bad Panda"]
144.217.50.240,["killer bear","LSD kitten"]
```

The result is an alert section that looks like:

```JSON
  "alert": {
    "action": "allowed",
    "gid": 1,
    "signature_id": 1,
    "rev": 0,
    "signature": "Test dataset",
    "category": "",
    "severity": 3,
    "extra": {
      "inventory": {
        "user": "vjulien",
        "rank": 1
      },
      "bad_actors": [
        "Killer Bear",
        "LSD kitten"
      ]
    }
  },
```

## Contribution style:
- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html

## Our Contribution agreements:
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)

## Changes (if applicable):
- [x] I have updated the User Guide (in [doc/userguide/](https://github.com/OISF/suricata/tree/304271e63a9e388412f25f0f94a1a0da4bf619d9/doc/userguide)) to reflect the changes made
- [x] I have updated the JSON schema (in [etc/schema.json](https://github.com/OISF/suricata/blob/304271e63a9e388412f25f0f94a1a0da4bf619d9/etc/schema.json)) to reflect all logging changes
      (including schema descriptions)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues

Link to ticket: https://redmine.openinfosecfoundation.org/issues/7372

Describe changes:
- Implement datajson keywords for all dataset types
- Add documentation
- Add unix socket commands

SV_REPO=https://github.com/regit/suricata-verify/
SV_BRANCH=datajson-v1.0

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .uptime | 636 | 676 | 106.29% | 

Pipeline 23289

File: src/datasets-json.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * \file
+ *
+ * \author Eric Leblond <el@stamus-networks.com>
+ */
+
+#ifndef __DATASETS_JSON_H__
+#define __DATASETS_JSON_H__
+
+#include <stdint.h>
```
Comment (victorjulien) [inline src/datasets-json.h:27]: should use suricata-common.h

File: src/datasets.c [hunk 2]
```diff
@@ -165,6 +170,22 @@ static int ParseRepLine(const char *in, size_t ins, DataRepType *rep_out)
     return 0;
 }
 
+static int ParseJsonLine(const char *in, size_t ins, DataJsonType *rep_out)
+{
+    json_error_t jerror;
+    json_t *msg;
+    msg = json_loads(in, 0, &jerror);
```
Comment (victorjulien) [inline src/datasets.c:177]: declare here

File: src/datasets.c [hunk 3]
```diff
@@ -165,6 +170,22 @@ static int ParseRepLine(const char *in, size_t ins, DataRepType *rep_out)
...
+    if (msg == NULL) {
+        SCLogWarning("dataset: Invalid json: %s\n", jerror.text);
+        return -1;
+    } else {
```
Comment (victorjulien) [inline src/datasets.c:181]: unnecessary else branch, if branch is terminating

File: src/datasets.c [hunk 4]
```diff
@@ -165,6 +170,22 @@ static int ParseRepLine(const char *in, size_t ins, DataRepType *rep_out)
...
+        json_decref(msg);
+    }
+    rep_out->len = ins;
+    rep_out->value = SCStrndup(in, ins);
```
Comment (victorjulien) [inline src/datasets.c:185]: need an error check here

File: src/datasets.c [hunk 5]
```diff
@@ -1224,6 +1539,30 @@ static DataRepResultType DatasetLookupIPv6wRep(
...
 
+static DataJsonResultType DatasetLookupIPv6wJson(
+        Dataset *set, const uint8_t *data, const uint32_t data_len)
+{
+    DataJsonResultType rrep = { .found = false, .json = { .value = NULL, .len = 0 } };
+
+    if (set == NULL)
+        return rrep;
+
+    if (data_len != 16 && data_len != 4)
```
Comment (victorjulien) [inline src/datasets.c:1550]: how would we get here with data_len == 4?

File: src/detect.h [hunk 6]
```diff
@@ -40,6 +40,7 @@
 #include "util-radix-tree.h"
 #include "util-file.h"
 #include "reputation.h"
+#include <stddef.h>
```
Comment (victorjulien) [inline src/detect.h:43]: suricata-common.h should be used

**Review by victorjulien — CHANGES_REQUESTED** [review]
Some minor comments.

Bigger issue: I like this work, but I would to understand why this needs a new set type and keyword type? Can we not overload the existing dataset/datarep facilities?

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 139 | 144 | 103.6% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 636 | 670 | 105.35% | 

Pipeline 23316

**Comment by catenacyber** [comment]
Did you forget to use https://github.com/OISF/suricata-verify/pull/2123 here ?

File: etc/schema.json [hunk 7]
```diff
@@ -212,6 +212,10 @@
                 "xff": {
                     "type": "string"
                 },
+                "extra": {
+                    "type": "object",
```
Comment (jasonish) [inline etc/schema.json:216]: I'm trying to get a description on everything new for documentation purposes, can you add one?

**Comment by jasonish** [comment]
I'm a little hung up on the "extra" object? Is this data much different than metadata? For example, if we were to spin a rule for each entry in dataset, we'd add extra context in the message, or metadata, here we're just adding extra metadata from the dataset.

Also, with "extra" being a very generic variable, should it be available to possible future items that might want to add extra metadata? For example, a Lua rule that wants to add extra metadata? Or should this be restricted to datajson and such be given a more specific name?

File: doc/userguide/rules/datasets.rst [hunk 8]
```diff
@@ -292,6 +350,17 @@ Syntax::
 
     <data>,<value>
 
+
+datajson
+~~~~~~~~
+
+The datajson format follows the dataset, expect that there are 1 more CSV
```
Comment (jasonish) [inline doc/userguide/rules/datasets.rst:357]: `expect` -> `except`?

**Comment by jasonish** [comment]
Is there a real need for a new dataset type of `datajson`? Could there just be `dataset`, and then `dataset` that has metadata?

File: src/datasets.c (hunk 5 above)
Comment (regit) [reply src/datasets.c:1550]: GetDataDst and GetDataSrc  in detect-ipaddr.c are setting a buffer length of 4 or 16 depending of IP type:
```C
static InspectionBuffer *GetDataDst(DetectEngineThreadCtx *det_ctx,
        const DetectEngineTransforms *transforms, Packet *p, const int list_id)
{
    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
    if (buffer->inspect == NULL) {
        if (PacketIsIPv4(p)) {
            /* Suricata stores the IPv4 at the beginning of the field */
            InspectionBufferSetup(det_ctx, list_id, buffer, p->dst.address.address_un_data8, 4);
        } else if (PacketIsIPv6(p)) {
            InspectionBufferSetup(det_ctx, list_id, buffer, p->dst.address.address_un_data8, 16);
```
so we can have the 2 lengths. As Suricata stores the IPv4 at the beginning of the field, this is working.

**Comment by regit** [comment]
> I'm a little hung up on the "extra" object? Is this data much different than metadata? For example, if we were to spin a rule for each entry in dataset, we'd add extra context in the message, or metadata, here we're just adding extra metadata from the dataset.
> 

I will admit I was kinda expecting to have a discussion on the used keyword. I'm not super happy with the choice but I did not find a better one.

I think we need to avoid conflict, people parsing metadata of signatures are kind of expecting to just have that in `metadata`. So it should be on its own.

> Also, with "extra" being a very generic variable name, should it be available to possible future items that might want to add extra metadata? For example, a Lua rule that wants to add extra metadata? Or should this be restricted to datajson and such be given a more specific name?

The concept of extracting information at detection time seems pretty good. Flowbits is currently used for that (lua, pcre extraction) but it is then propagated to all events in the flowbits object. It may be better to offer a per alert way of doing this.

Maybe we can stay with `extra` (or something generic) in this case and extend the capabilities to other systems.

**Comment by jasonish** [comment]
> > Is there a real need for a new dataset type of `datajson`? Could there just be `dataset`, and then `dataset` that has metadata?
> 
> I think it is more on educational side. Potentially also a bit keeping the code simple. On the educational/usage side, it allows signature readers to understand really fast what is really done. datajson keyword is rather explicit and different from dataset so user can understand that there will be consequences.

Can you explain how its different than a dataset that contains additional metadata to attach to the alert?  Its almost like `datadata` would be the correct keyword then.  JSON is just the encoding of the `data` :)

File: doc/userguide/rules/datasets.rst [hunk 9]
```diff
@@ -292,6 +350,17 @@ Syntax::
...
+field:
+
+Syntax::
+
+    <data>,<json_data>
```
Comment (jasonish) [inline doc/userguide/rules/datasets.rst:362]: Can you add an example with simple JSON data?  Is it still a valid CSV?

File: src/detect-datajson.c [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+        const uint8_t *data, const uint32_t data_len)
+{
+    if (data == NULL || data_len == 0)
+        return 0;
+
+    switch (sd->cmd) {
+        case DETECT_DATAJSON_CMD_ISSET: {
+            // PrintRawDataFp(stdout, data, data_len);
+            DataJsonResultType r = DatasetLookupwJson(sd->set, data, data_len);
+            SCLogDebug("r %d", r);
```
Comment (jasonish) [inline src/detect-datajson.c:69]: Fails with --enable-debug:
```
detect-datajson.c:69:24: warning: format ‘%d’ expects argument of type ‘int’, but argument 7 has type ‘DataJsonResultType’ [-Wformat=]
```

**Comment by regit** [comment]
Closing in favor of #12175

**Pull request closed without merging**